File: java/triangle/src/main/java/Triangle.java
import java.util.stream.DoubleStream;

class Triangle {
    private final double[] sides;

    Triangle(double side1, double side2, double side3) throws TriangleException {
        this.sides = DoubleStream.of(side1, side2, side3).sorted().toArray();
        if (sides[0] <= 0.0 || (sides[0] + sides[1]) < sides[2]) {
            throw new TriangleException();
        }
    }

    boolean isEquilateral() {
        return sides[0] == sides[1] && sides[1] == sides[2];
    }

    boolean isIsosceles() {
        return sides[0] == sides[1] || sides[1] == sides[2];
    }

    boolean isScalene() {
        return sides[0] != sides[1] && sides[1] != sides[2];
    }
}

File: typescript/triangle/triangle.ts
type TriangleKind = 'invalid' | 'scalene' | 'isosceles' | 'equilateral';

const uniqueSidesToKind: Record<number, TriangleKind> = {
  1: 'equilateral',
  2: 'isosceles',
  3: 'scalene',
};

export class Triangle {
  private readonly kind: TriangleKind;

  constructor(...sides: number[]) {
    const sortedSides = [...sides].sort((a, b) => a - b);

    if (sortedSides[0] + sortedSides[1] >= sortedSides[2] && sortedSides[0] > 0) {
      this.kind = uniqueSidesToKind[new Set<number>(sortedSides).size];
    } else {
      this.kind = 'invalid';
    }
  }

  get isEquilateral(): boolean {
    return this.kind === 'equilateral';
  }

  get isIsosceles(): boolean {
    return this.isEquilateral || this.kind === 'isosceles';
  }

  get isScalene(): boolean {
    return this.kind === 'scalene';
  }
}
